etcdserver: exclude learner from leader transfer

jingyih opened the pull request:
This PR does:
1. Maintenance API MoveLeader() returns `ErrBadLeaderTransferee` if the leadership transferee does not exist or is raft learner.

```
$ etcdctl member list -w table
+------------------+---------+--------+------------------------+------------------------+------------+
|        ID        | STATUS  |  NAME  |       PEER ADDRS       |      CLIENT ADDRS      | IS LEARNER |
+------------------+---------+--------+------------------------+------------------------+------------+
| 8211f1d0f64f3269 | started | infra1 | http://127.0.0.1:12380 |  http://127.0.0.1:2379 |      false |
| ca54cf488514d845 | started | infra2 | http://127.0.0.1:22380 | http://127.0.0.1:22379 |       true |
+------------------+---------+--------+------------------------+------------------------+------------+

$ etcdctl move-leader ca54cf488514d845
{"level":"warn","ts":"2019-04-12T17:32:46.189-0700","caller":"clientv3/retry_interceptor.go:60","msg":"retrying of unary invoker failed","target":"endpoint://client-786c46e9-8602-464f-a449-b6991ee4bff6/127.0.0.1:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: bad leader transferee"}
Error: etcdserver: bad leader transferee
```

2. etcdserver `TransferLeadership()` **ONLY** choose voting member as transferee. When etcdserver is been signaled to stop, it will transfer its leadership to another member in the cluster by calling `TransferLeadership()`. 

Without this change, learner member might be picked as leadership transferee. The leadership transfer operation will timeout, as the request is dropped silently in the raft layer.
https://github.com/jingyih/etcd/blob/e1ca3b4434945e57e8e3a451cdbde74a903cc8e1/raft/raft.go#L1168-L1172

jingyih committed:
etcdserver: exclude learner from leader transfer

1. Maintenance API MoveLeader() returns ErrBadLeaderTransferee if
transferee does not exist or is raft learner.

2. etcdserver TransferLeadership() only choose voting member as
transferee.

jingyih commented:
/cc @gyuho @WIZARD-CXY

WIZARD-CXY commented on etcdserver/server.go:
```diff
@@ -1435,20 +1435,20 @@ func (s *EtcdServer) TransferLeadership() error {
 		return nil
 	}
 
-	if !s.isMultiNode() {
```
can we remove `isMultiNode` or just encapsulate `if s.cluster == nil || len(s.cluster.VotingMemberIDs()) <= 1 { ` in it?

jingyih replied:
It is already removed.

WIZARD-CXY commented:
can we add a test case in the `server_test.go` for this MoveLeader case?

jingyih commented:
> can we add a test case in the `server_test.go` for this MoveLeader case?

Good idea. I'll take a look.

WIZARD-CXY commented:
Overall LGTM

jingyih commented:
I'll add integration test in the future PR. It is not obvious to me how to do meaningful unit test regarding leadership transfer.

jingyih commented:
cc @jpbetz

jingyih committed:
integration: add TestMoveLeaderToLearnerError

Adding integration test TestMoveLeaderToLearnerError, which ensures that
leader transfer to learner member will fail.

jingyih commented:
Added an integration test to ensure leader transfer to learner member will fail.

jingyih committed:
integration: add TestTransferLeadershipWithLearner

Adding integration test TestTransferLeadershipWithLearner, which ensures
that TransferLeadership does not timeout due to learner is automatically
picked by leader as transferee.

jingyih commented:
Added another integration test to ensure automatically leadership transfer does not pick learner as transferee, which will result in leadership transfer timeout.
